File: AULA 10/E - Exchange Rates.java
import java.util.*;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Map<String, Map<String, Pair<Integer, Integer>>> relations = new HashMap<>();

        while (true) {
            String[] input = scanner.nextLine().split(" ");

            if (input[0].equals("!")) {
                int m = Integer.parseInt(input[1]);
                int n = Integer.parseInt(input[4]);

                String itema = input[2].trim();
                String itemb = input[5].trim();

                int factor = GCD(m, n);

                m /= factor;
                n /= factor;

                relations.computeIfAbsent(itema, k -> new HashMap<>()).put(itemb, new Pair<>(m, n));
                relations.computeIfAbsent(itemb, k -> new HashMap<>()).put(itema, new Pair<>(n, m));
                
            } else if (input[0].equals("?")) {
                String itema = input[1].trim();
                String itemb = input[3].trim();

                Pair<Boolean, Pair<Integer, Integer>> result = DFS(relations, itema, itemb);

                String strResult = (result.getKey() ? result.getValue().getKey() : "?") + " " + itema + " = " + (result.getKey() ? result.getValue().getValue() : "?") + " " + itemb;
                System.out.println(strResult);
                
            } else {
                break;
            }
        }
        scanner.close();
    }

    public static int GCD(int a, int b) {
        while (b != 0) {
            int r = a % b;
            a = b;
            b = r;
        }
        return a;
    }

    public static Pair<Boolean, Pair<Integer, Integer>> DFS(Map<String, Map<String, Pair<Integer, Integer>>> relations, String itema, String itemb) {
        Map<String, Boolean> visited = new HashMap<>();
        for (String key : relations.keySet()) {
            visited.put(key, false);
        }
        return DFS(relations, visited, itema, itemb);
    }

    public static Pair<Boolean, Pair<Integer, Integer>> DFS(Map<String, Map<String, Pair<Integer, Integer>>> relations, Map<String, Boolean> visited, String itema, String itemb) {
        visited.put(itema, true);

        for (Map.Entry<String, Pair<Integer, Integer>> rel : relations.getOrDefault(itema, Collections.emptyMap()).entrySet()) {
            if (rel.getKey().equals(itemb)) {
                return new Pair<>(true, rel.getValue());
            } else if (!visited.get(rel.getKey())) {
                Pair<Boolean, Pair<Integer, Integer>> result = DFS(relations, visited, rel.getKey(), itemb);
                if (result.getKey()) {
                    int x = result.getValue().getKey() * rel.getValue().getKey();
                    int y = result.getValue().getValue() * rel.getValue().getValue();

                    int factor = GCD(x, y);

                    x /= factor;
                    y /= factor;

                    return new Pair<>(true, new Pair<>(x, y));
                }
            }
        }

        return new Pair<>(false, new Pair<>(0, 0));
    }

    static class Pair<K, V> {
        private final K key;
        private final V value;

        public Pair(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}

File: AULA 10/C#/E - Exchange Rates.cs
using System;
using System.Text;
using System.Collections.Generic;

public static class Program
{
    public static void Main()
    {
        var relations = new Dictionary<string, Dictionary<string, Tuple<int, int>>>();

        while (true)
        {
            var input = Console.ReadLine()!.Split(' ');

            if (input[0] == "!")
            {
                var m = int.Parse(input[1]);
                var n = int.Parse(input[4]);

                var itema = input[2].Trim();
                var itemb = input[5].Trim();

                var factor = GCD(m, n);

                m /= factor;
                n /= factor;

                var relationA = new Dictionary<string, Tuple<int, int>>
                {
                    { itemb, new Tuple<int, int>(m, n) }
                };

                var relationB = new Dictionary<string, Tuple<int, int>>
                {
                    { itema, new Tuple<int, int>(n, m) }
                };

                if(!relations.ContainsKey(itema))
                {
                    relations.TryAdd(itema, relationA);
                }
                else if(!relations[itema].ContainsKey(itemb))
                {
                    relations[itema].TryAdd(itemb, new Tuple<int, int>(m, n));
                }

                if (!relations.ContainsKey(itemb))
                {
                    relations.TryAdd(itemb, relationB);
                }
                else if (!relations[itemb].ContainsKey(itema))
                {
                    relations[itemb].TryAdd(itema, new Tuple<int, int>(n, m));
                }
            }
            else if (input[0] == "?")
            {
                var itema = input[1].Trim();
                var itemb = input[3].Trim();

                var result = DFS(relations, itema, itemb);

                var strResult = $"{(result.Item1 ? $"{result.Item2.Item1}" : "?")} {itema} = {(result.Item1 ? $"{result.Item2.Item2}" : "?")} {itemb}";

                Console.WriteLine(strResult);
            }
            else
            {
                break;
            }
        }
    }

    public static int GCD(int a, int b)
    {
        while (b != 0)
        {
            int r = a % b;
            a = b;
            b = r;
        }

        return a;
    }

    public static Tuple<bool, Tuple<int, int>> DFS(Dictionary<string, Dictionary<string, Tuple<int, int>>> relations, string itema, string itemb)
    {
        var visited = new Dictionary<string, bool>();

        foreach (var rel in relations)
        {
            visited.TryAdd(rel.Key, false);
        }

        return DFS(relations, visited, itema, itemb);
    }

    public static Tuple<bool, Tuple<int, int>> DFS(Dictionary<string, Dictionary<string, Tuple<int, int>>> relations, Dictionary<string, bool> visited, string itema, string itemb)
    {
        visited[itema] = true;

        foreach (var rel in relations[itema])
        {
            if(rel.Key == itemb)
            {
                return Tuple.Create(true, rel.Value);
            }
            else if (!visited[rel.Key])
            {
                var next = rel.Key;

                var result = DFS(relations, visited, next, itemb);

                if(result.Item1)
                {
                    var x = result.Item2.Item1 * rel.Value.Item1;
                    var y = result.Item2.Item2 * rel.Value.Item2;

                    var factor = GCD(x, y);

                    x /= factor;
                    y /= factor;

                    return Tuple.Create(true, Tuple.Create(x, y));
                }
            }
        }

        return Tuple.Create(false, Tuple.Create(0, 0));
    }
}